keter.nix

```diff
@@ -0,0 +1,71 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.services.keter-ng;
+
+in
+{
+  options.services.keter-ng = {
+    enable = lib.mkEnableOption (lib.mdDoc ''
+      keter — a web app deployment manager.
+    '');
+
+    root = lib.mkOption {
+      type = lib.types.str;
+      default = "/opt/keter";
+      description = lib.mdDoc "Mutable state folder for keter";
+    };
+
+    package = lib.mkOption {
+      type = lib.types.package;
+      default = pkgs.haskellPackages.keter;
+      defaultText = lib.literalExpression "pkgs.haskellPackages.keter";
+      description = lib.mdDoc "The keter package to be used";
+    };
+
+    globalKeterConfig = lib.mkOption {
+      type = lib.types.str;
+      default = "";
+      description = lib.mdDoc ''
+        A custom YAML configuration for Keter. This content will be directly
+        written to the `keter-config.yml` file without modification. See
+        <https://github.com/snoyberg/keter/blob/master/etc/keter-config.yaml>
+        for reference.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable (
+    let
+      incoming = "${cfg.root}/incoming";
+
+      globalKeterConfigFile = pkgs.writeTextFile {
+        name = "keter-config.yml";
+        text = cfg.globalKeterConfig;
+      };
+    in
+    {
+      systemd.services.keter-ng = {
+        description = "keter app loader";
+        script = ''
+          set -xe
+          mkdir -p ${incoming}
+          ${lib.getExe cfg.package} ${globalKeterConfigFile};
+        '';
+        wantedBy = [ "multi-user.target" "nginx.service" ];
+
+        serviceConfig = {
+          Restart = "always";
+          RestartSec = "10s";
+        };
+
+        after = [
+          "network.target"
+          "local-fs.target"
+          "postgresql.service"
+        ];
+      };
+    }
+  );
+}
```

Comment:
fyi,

keter is upstreamed as a module in nixpkgs:
https://github.com/NixOS/nixpkgs/blob/335958709c637736c44fdfde6f8b7ca5c6bc8da0/nixos/modules/services/web-servers/keter/default.nix#L4

Reply:
I did see that, but I don't think the module definition is quite right for a couple of reasons. From memory,

1. It doesn't support all of the possible configuration options for the global configuration
2. It concerns itself with the bundle configuration, when it shouldn't

The comment describing the indirection is already a good hint that this is trying to do too much. The easiest thing to do is to delegate configuring this correctly to the user.